src/main.rs

```diff
@@ -0,0 +1,116 @@
+use std::{
+    env, fs,
+    io::{stderr, Write},
+    path::Path,
+    process,
+};
+
+use getopts::Options;
+use libc::{c_int, getchar, putchar};
+
+fn usage(prog: &str) -> ! {
+    let path = Path::new(prog);
+    let leaf = path
+        .file_name()
+        .map(|x| x.to_str().unwrap_or("ykbf"))
+        .unwrap_or("ykbf");
+    writeln!(&mut stderr(), "Usage: {} [-h] <file.bf>", leaf).ok();
+    process::exit(1)
+}
+
+fn compile(txt: &str) -> (String, Vec<usize>) {
+    // BF comments are anything that's not a command character, so scrub all non-command
+    // characters.
+    let mut out = String::with_capacity(txt.len());
+    for c in txt.chars() {
+        match c {
+            '>' | '<' | '+' | '-' | '.' | ',' | '[' | ']' => {
+                out.push(c);
+            }
+            _ => (),
+        }
+    }
+    out.shrink_to_fit();
+
+    // Pre-calculate the [...] jumps as a map from PC -> PC.
+    let mut map = Vec::with_capacity(out.len());
+    map.resize(out.len(), 0);
+    let mut stack = Vec::new();
+    for (i, c) in out.chars().enumerate() {
+        if c == '[' {
+            stack.push(i);
+        } else if c == ']' {
+            let s = stack.pop().unwrap();
+            map[s] = i;
+            map[i] = s;
+        }
+    }
+
+    (out, map)
+}
+
+fn interp(prg: &[u8], map: Vec<usize>) {
+    let mut pc = 0;
+    let mut ptr = 0;
+    let mut store = Vec::with_capacity(30000);
+    store.resize(30000, 0u8);
+    while pc < prg.len() {
+        match prg[pc] as char {
+            '>' => {
+                if ptr == store.len() {
+                    store.push(0);
+                }
+                ptr += 1;
+            }
+            '<' => {
+                if ptr > 0 {
+                    ptr -= 1;
+                }
+            }
+            '+' => {
+                store[ptr] = store[ptr].wrapping_add(1);
+            }
+            '-' => {
+                store[ptr] = store[ptr].wrapping_sub(1);
+            }
+            '.' => {
+                let b = store[ptr];
+                unsafe {
+                    putchar(b as c_int);
+                }
+            }
+            ',' => {
+                store[ptr] = unsafe { getchar() } as u8;
+            }
+            '[' => {
+                if store[ptr] == 0 {
+                    pc = map[pc];
+                }
+            }
+            ']' => {
+                if store[ptr] != 0 {
+                    pc = map[pc];
+                }
+            }
+            _ => unreachable!(),
+        }
+        pc += 1;
+    }
+}
+
+fn main() {
+    let args: Vec<String> = env::args().collect();
+    let prog = &args[0];
+    let matches = Options::new()
+        .optflag("h", "help", "")
+        .parse(&args[1..])
+        .unwrap_or_else(|_| usage(prog));
+    if matches.opt_present("h") || matches.free.len() != 1 {
+        usage(prog);
+    }
+
+    let inv = fs::read(&matches.free[0]).expect("Can't read file.");
```

Comment:
A nitpick, but what does `inv` stand for here? It's the input bytes, right?

Reply:
What would you call it? It's a minor temporary variable: its name doesn't really seem *that* important to me.

Reply:
It's not that important at all. I'd have called it `bytes` perhaps.